Fixes #307: Finish cellular functionality for audio streaming

jamesxu0 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #307: Finish cellular functionality for audio streaming. I moved the audio test cases from StateFragmentTest to ExplorationActivityTest. Also, I removed CellularDataDialogFragment since the audio button was moved to ExplorationActivity. This PR also adds NetworkConnectionUtil to get the current network status.

ExplorationInjectionActivity is used to get references to certain dependencies that the application uses. In this case, we need ExplorationDataController to start the playing the exploration before opening the ExplorationActivity. Also, we need to set the network status in NetworkConnectionUtil.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

jamesxu0 committed:
Finished audio functionality for statefragment

jamesxu0 committed:
moved audio fragment to top

jamesxu0 committed:
Added test cases to NetworkConnectionUtil

jamesxu0 committed:
Added remove audiofragment animation

jamesxu0 committed:
moved exploration retrieval to AudioViewModel

jamesxu0 committed:
Added no audio available

jamesxu0 committed:
Changed NetworkConnectionUtil for tests

jamesxu0 committed:
new audio test cases, removed old ones

jamesxu0 committed:
Merge branch 'develop' into state-fragment-audio

jamesxu0 committed:
Fixed tests

jamesxu0 committed:
Fixed comment

jamesxu0 committed:
Fixed rotation case

jamesxu0 committed:
Removed unused methods and imports

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -1,27 +1,43 @@
+
+  var selectedLanguageCode: String = ""
+  var languages = listOf<String>()
+  var isLanguagesEmpty = ObservableField(false)
```
Why not make languages an observable field and check that, instead?

jamesxu0 replied:
Updated languages to an ObservableField and removed isLanguagesEmpty.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -44,16 +60,41 @@ class AudioViewModel @Inject constructor(
-    voiceoverMap = map
+  val showSeekBar = ObservableField(true)
+
+  fun getExplorationById(exploreId: String, stateId: String) {
```
Regarding the name, isn’t returning anything. Maybe this should be initializeVoiceoverMappings, or similar?

jamesxu0 replied:
Updated to initializeVoiceoverMappings.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -44,16 +60,41 @@ class AudioViewModel @Inject constructor(
 
-  fun setExplorationId(id: String) {
-    explorationId = id
+  fun setVoiceoverMappingsByState(stateId: String) {
```
Can this be private? Prefer minimizing visibility.

jamesxu0 replied:
No because this function is used in StateFragmentPresenter.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -44,16 +60,41 @@ class AudioViewModel @Inject constructor(
+    } else if (languages.any { it == selectedLanguageCode }) {
+      setAudioLanguageCode(selectedLanguageCode)
+    } else {
+      if (selectedLanguageCode.isEmpty() && languages.isNotEmpty()) {
```
Nit: we can merge this with the else above.

jamesxu0 replied:
Updated to a when statement.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -64,4 +65,7 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterf
     return super.onOptionsItemSelected(item)
   }
+
+  fun showVolumeOff() = explorationActivityPresenter.showVolumeOff()
```
For this and next file, it might be clearer to refer to audio streaming rather than volume since this button doesn’t actually correspond to volume.

jamesxu0 replied:
Updated names to showAudioStreamingOn/Off.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -44,6 +45,15 @@ class ExplorationActivityPresenter @Inject constructor(
+  }
+
+  /** Sets audio button drawable to volume off. */
+  fun showVolumeOff() = audioButton.setIcon(R.drawable.ic_volume_off_48dp)
```
Could we set this drawable via databinding instead?

jamesxu0 replied:
Updated to use databinding.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -44,16 +60,41 @@ class AudioViewModel @Inject constructor(
+    } else {
+      if (selectedLanguageCode.isEmpty() && languages.isNotEmpty()) {
+        setAudioLanguageCode(languages.first())
+        (fragment as AudioFragment).languageSelectionClicked()
```
Let’s use an interface to avoid a direct dependency on AudioFragment.

jamesxu0 replied:
Updated to cast to LanguageInterface.

BenHenning commented on app/src/main/java/org/oppia/app/testing/ExplorationInjectionActivity.kt:
```diff
@@ -0,0 +1,18 @@
+
+  fun getExplorationDataController() = explorationInjectionActivityPresenter.explorationDataController
+  fun getNetworkConnectionUtil() = explorationInjectionActivityPresenter.networkConnectionUtil
+}
```
Nit: new lines at end of all files.

jamesxu0 replied:
Fixed.

BenHenning commented on app/src/main/java/org/oppia/app/testing/ExplorationInjectionActivity.kt:
```diff
@@ -0,0 +1,18 @@
+    activityComponent.inject(this)
+  }
+
+  fun getExplorationDataController() = explorationInjectionActivityPresenter.explorationDataController
```
We don’t actually need a presenter. We can just inject these directly via lateinit vars and access them directly. This is fine for testing purposes.

jamesxu0 replied:
Removed presenter and updated to directly inject activity.

BenHenning commented on app/src/main/java/org/oppia/app/testing/ExplorationInjectionActivity.kt:
```diff
@@ -0,0 +1,18 @@
+import org.oppia.app.activity.InjectableAppCompatActivity
+import javax.inject.Inject
+
+class ExplorationInjectionActivity : InjectableAppCompatActivity() {
```
Nit: please add kdoc.

jamesxu0 replied:
Added kdoc.

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -8,6 +8,10 @@
   <string name="audio_language_select_dialog_okay_button">OK</string>
   <string name="audio_language_select_dialog_cancel_button">Cancel</string>
   <string name="audio_language_select_dialog_title">Audio Language</string>
+  <string name="audio_no_audio_available">No audio available.</string>
```
Are these strings signed off by Chantel?

jamesxu0 replied:
Yes, the strings match the designs.

BenHenning commented on domain/src/main/assets/fractions_exploration0.json:
```diff
@@ -724,7 +724,7 @@
         "voiceovers_mapping": {
           "content": {
-            "en": {
+            "hi-en": {
```
Why these changes? Do we still need these after the recent updates to these JSON files?

jamesxu0 replied:
These changes were just for testing. Reverted back to original.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -243,23 +275,34 @@ class StateFragmentPresenter @Inject constructor(
   }
 
   private fun showHideAudioFragment(isVisible: Boolean) {
```
Perhaps: setAudioFragmentVisible would work a bit better with the boolean.

jamesxu0 replied:
Updated name to setAudioFragmentVisible.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -243,23 +275,34 @@ class StateFragmentPresenter @Inject constructor(
-        fragment.childFragmentManager.beginTransaction().remove(getAudioFragment()!!).commitNow()
+      (fragment.requireActivity() as ExplorationActivity).showVolumeOff()
+      if (audioFragment != null) {
+        val animation = AnimationUtils.loadAnimation(context, R.anim.slide_up_audio)
```
I’m wondering: could we always add the fragment and instead animate the parent view and set it’s visibility based on enabled state? That might be easier and a bit safer than relying on a callback to perform a fragment transaction.

jamesxu0 replied:
Updated so the audio fragment is always added but it's visibility is data bound.

BenHenning commented on app/src/main/res/drawable/ic_volume_on_48dp.xml:
```diff
@@ -0,0 +1,5 @@
+<vector android:autoMirrored="true" android:height="48dp"
```
Something I noticed this weekend: I think we want to import 24dp audio icons and render at 24, but have a 48dp touch target. The audio icon looked a bit large to me in the app.

jamesxu0 replied:
Updated icons to 24dp.

BenHenning commented on app/src/main/res/layout/audio_fragment.xml:
```diff
@@ -50,23 +47,39 @@
       android:thumb="@drawable/seekbar_thumb"
+      android:visibility="@{viewModel.showSeekBar ? View.VISIBLE : View.GONE}"
+      android:layout_marginStart="15dp"
+      android:thumbOffset="15dp"
```
What’s this for?

jamesxu0 replied:
This is used to move the thumb of the seekbar to the start.

BenHenning commented on app/src/main/res/layout/audio_fragment.xml:
```diff
@@ -50,23 +47,39 @@
       app:layout_constraintEnd_toStartOf="@+id/tvAudioLanguage"
       app:layout_constraintStart_toEndOf="@+id/ivPlayPauseAudio"
       app:layout_constraintTop_toTopOf="parent" />
 
```
Can you add screenshots to the PR showing each of the different possible states the audio bar and dialog can be in? It’d be good to see.

jamesxu0 replied:
I added a comment to include screenshots of the audio bar and dialogs it opens.

BenHenning commented on domain/src/main/java/org/oppia/domain/exploration/ExplorationRetriever.kt:
```diff
@@ -33,7 +33,7 @@ class ExplorationRetriever @Inject constructor(
       TEST_EXPLORATION_ID_5 -> loadExplorationFromAsset("welcome.json")
       TEST_EXPLORATION_ID_6 -> loadExplorationFromAsset("about_oppia.json")
-      TEST_EXPLORATION_ID_30 -> loadExplorationFromAsset("prototype_exploration.json")
+      TEST_EXPLORATION_ID_30 -> loadExplorationFromAsset("fractions_exploration0.json")
```
We shouldn't be changing this ID association. Could you use FRACTIONS_EXPLORATION_ID_0, instead, if you want to load that?

jamesxu0 replied:
This was an accidental change. Reverted back to original.

BenHenning commented on domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt:
```diff
@@ -114,12 +115,20 @@ class AudioPlayerController @Inject constructor(
-        AsyncResult.success(PlayProgress(PlayStatus.PREPARED, 0, it.duration))
+        AsyncResult.success(PlayProgress(PlayStatus.PREPARED, 0, duration))
+    }
+    mediaPlayer.setOnErrorListener {_, what, _ ->
```
Should we always release if there's an error? What circumstances warrant keeping the media player active and having an error?

jamesxu0 replied:
I think the media player should always be released on error. Updated to always release and initialize on error.

BenHenning commented on domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt:
```diff
@@ -114,12 +115,20 @@ class AudioPlayerController @Inject constructor(
+        releaseMediaPlayer()
+        initializeMediaPlayer()
+      }
+      true
```
What does this return value correspond to in the error listener? It might be nice to have a brief comment justifying the 'true' value being provided here.

jamesxu0 replied:
This return value lets the listener know if the error has been handled and whether to call on completion.

BenHenning commented on domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt:
```diff
@@ -114,12 +115,20 @@ class AudioPlayerController @Inject constructor(
-        AsyncResult.success(PlayProgress(PlayStatus.PREPARED, 0, it.duration))
+        AsyncResult.success(PlayProgress(PlayStatus.PREPARED, 0, duration))
+    }
+    mediaPlayer.setOnErrorListener {_, what, _ ->
```
Are there any behavioral differences here that we should add tests for? What breaks if this regresses?

jamesxu0 replied:
I added a test to AudioPlayerControllerTest to ensure the OnErrorListener works correctly.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -42,16 +67,115 @@ class ExplorationActivityTest {
+    ActivityScenario.launch<ExplorationActivity>(createExplorationActivityIntent(FRACTIONS_EXPLORATION_ID_0)).use {
+      onView(withId(R.id.action_audio_player)).perform(click())
+      onView(withId(R.id.continue_button)).perform(click())
+      onView(withText(context.getString(R.string.audio_no_audio_available))).check(matches(isDisplayed()))
```
I'm still not sure about this behavior. I kinda feel like we shouldn't show the audio icon at all if there isn't audio available for that lesson or state, or maybe show a dialog. This is worth opening a thread with Chantel to discuss our options.

Edit: can we update the behavior to match: https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/oppia-android-dev/eojdNU9w9ZA/PC4nPvIJAwAJ (including default selection)?

jamesxu0 replied:
Updated the behavior of the audio button to match the discussion.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -42,16 +67,115 @@ class ExplorationActivityTest {
+  }
+
+  @Test
+  fun testAudioWithCellular_clickAudioIcon_clickPositive_checkAudioFragment() {
```
Can we be more specified than 'checkAudioFragment'? Are we checking that it's visible? Ditto elsewhere for similarly named tests.

jamesxu0 replied:
Updated the test names to checkAudioFragmentIsHidden/Displayed.

BenHenning commented on utility/src/main/java/org/oppia/util/networking/NetworkConnectionUtil.kt:
```diff
@@ -0,0 +1,40 @@
+import javax.inject.Singleton
+
+@Singleton
+class NetworkConnectionUtil @Inject constructor(private val context: Context) {
```
Please add kdoc.

jamesxu0 replied:
Updated.

BenHenning commented on utility/src/main/java/org/oppia/util/networking/NetworkConnectionUtil.kt:
```diff
@@ -0,0 +1,40 @@
+
+@Singleton
+class NetworkConnectionUtil @Inject constructor(private val context: Context) {
+  enum class ConnectionStatus {
```
Also ditto here--let's specify what this is & what each of the constants are.

jamesxu0 replied:
Added comments for each constant.

BenHenning replied:
Also, please add a KDoc for the enum class itself.

jamesxu0 replied:
Added KDoc.

BenHenning commented on utility/src/main/java/org/oppia/util/networking/NetworkConnectionUtil.kt:
```diff
@@ -0,0 +1,40 @@
+  fun setCurrentConnectionStatus(status: ConnectionStatus) {
+    connectionStatus = status
+  }
+}
```
Add EOF newline.

jamesxu0 replied:
Updated.

BenHenning commented on utility/src/main/java/org/oppia/util/networking/NetworkConnectionUtil.kt:
```diff
@@ -0,0 +1,40 @@
+    val isConnected: Boolean = activeNetwork?.isConnected == true
+    val isWiFi: Boolean = activeNetwork?.type == ConnectivityManager.TYPE_WIFI
+    val isMob: Boolean = activeNetwork?.type == ConnectivityManager.TYPE_MOBILE
+    if(isConnected) {
```
Nit: spacing between ``if`` and ``(``. Also, I think we can simplify this a bit with a ``when`` statement:

```kotlin
val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
return connectivityManager.activeNetworkInfo?.let { activeNetwork ->
  val isConnected = activeNetwork.isConnected
  val isLocal = activeNetwork.type == ConnectivityManager.TYPE_WIFI || ConnectivityManager.TYPE_ETHERNET
  val isCellular = activeNetwork.type == ConnectivityManager.TYPE_MOBILE || ConnectivityManager.TYPE_WIMAX
  return@let when {
    isConnected && isLocal -> ConnectionStatus.LOCAL
    isConnected && isCellular -> ConnectionStatus.CELLULAR
    else -> ConnectionStatus.NONE
  }
} ?: ConnectionStatus.NONE
```

I also rearranged some things and renamed WIFI to LOCAL since Ethernet is definitely a valid transport type for streaming audio.

jamesxu0 replied:
Updated with this code block.

BenHenning commented on utility/src/main/java/org/oppia/util/networking/NetworkConnectionUtil.kt:
```diff
@@ -0,0 +1,40 @@
+  enum class ConnectionStatus {
+    WIFI, CELLULAR, NONE
+  }
+  private var connectionStatus: ConnectionStatus? = null
```
Perhaps ``testConnectionStatus`` to illustrate this isn't a cached value.

jamesxu0 replied:
Updated.

BenHenning commented on utility/src/test/java/org/oppia/util/networking/NetworkConnectionUtilTest.kt:
```diff
@@ -0,0 +1,161 @@
+  @Test
+  fun testGetCurrentConnectionStatus_setConnectivityManagerToWifi_checkConnectionStatus() {
+    shadowOf(context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager)
+      .setActiveNetworkInfo(ShadowNetworkInfo.newInstance(null, ConnectivityManager.TYPE_WIFI, 0, true, true))
```
Let's move the network info construction to a helper so that we can hide some of these values.

Also, please add argument comments for these literals.

jamesxu0 replied:
Moved into helper function and added argument comments.

BenHenning commented on utility/src/test/java/org/oppia/util/networking/NetworkConnectionUtilTest.kt:
```diff
@@ -0,0 +1,161 @@
+  }
+
+  @Test
+  fun testGetCurrentConnectionStatus_setConnectivityManagerToWifi_checkConnectionStatus() {
```
For naming, maybe something like:

``testGetCurrentCOnnectionStatus_activeWifiConnection_returnsWifi``

Ditto elsewhere.

jamesxu0 replied:
Updated.

BenHenning commented on utility/src/test/java/org/oppia/util/networking/NetworkConnectionUtilTest.kt:
```diff
@@ -0,0 +1,161 @@
+/** Tests for [DirectoryManagementUtil]. */
+@RunWith(AndroidJUnit4::class)
+@Config(manifest = Config.NONE)
+class NetworkConnectionUtilTest {
```
Other tests that might be nice:
- Wifi without connectivity
- No active network (since it can be null)
- Active ethernet
- Inactive ethernet
- Active wimax
- Inactive wimax
- Bluetooth (should not be recognized as a source for internet currently)

jamesxu0 replied:
Updated the test suite to include these tests.

BenHenning commented on utility/src/test/java/org/oppia/util/networking/NetworkConnectionUtilTest.kt:
```diff
@@ -0,0 +1,161 @@
+import javax.inject.Qualifier
+import javax.inject.Singleton
+
+/** Tests for [DirectoryManagementUtil]. */
```
Update comment.

jamesxu0 replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -44,16 +60,41 @@ class AudioViewModel @Inject constructor(
+      setAudioLanguageCode(selectedLanguageCode)
+    } else {
+      if (selectedLanguageCode.isEmpty() && languages.isNotEmpty()) {
+        setAudioLanguageCode(languages.first())
```
Per the email thread, shouldn't we select English by default if it's available, and then the first language otherwise?

jamesxu0 replied:
The current logic does try to find the default language first. languages.first() is only used if the default language and selected language are not present in the list.

BenHenning reviewed (changes requested):
Thanks @jamesxu0!

jamesxu0 committed:
Added feedback playback

jamesxu0 committed:
Added autoplay

jamesxu0 committed:
Always add audio fragment

jamesxu0 committed:
Working on test cases

jamesxu0 committed:
Added exploration tests and fixed not showing exploration title

jamesxu0 committed:
Added more test cases

jamesxu0 committed:
Do not auto play audio when language selection is shown

jamesxu0 commented:
Here are screenshots of the audio bar and the dialogs it opens:

**Audio Bar**
![Screenshot_20191121-112321](https://user-images.githubusercontent.com/40580399/69383364-4e6e2a80-0c6e-11ea-8d2b-9a42194b5b50.jpg)

**Offline Dialog**
![Screenshot_20191121-112350](https://user-images.githubusercontent.com/40580399/69383402-6180fa80-0c6e-11ea-875b-f9ad9971268e.jpg)

**Cellular Data Warning Dialog**
![Screenshot_20191121-112308](https://user-images.githubusercontent.com/40580399/69383415-6ba2f900-0c6e-11ea-8a59-c4dc7384d59d.jpg)

**Language Selection Dialog**
![Screenshot_20191121-112330](https://user-images.githubusercontent.com/40580399/69383431-7493ca80-0c6e-11ea-9022-89629ecaac81.jpg)

jamesxu0 committed:
Fixed test case

jamesxu0 committed:
Merge branch 'develop' into state-fragment-audio

jamesxu0 commented:
@BenHenning PTAL. I addressed all comments, fixed functionality to match specs, added auto play, and wrote more test cases.

jamesxu0 committed:
Merge branch 'develop' into state-fragment-audio

jamesxu0 committed:
Merge branch 'develop' into state-fragment-audio

rt4914 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -5,6 +5,7 @@
   <!-- TODO(#56): Reenable landscape support. -->
 
   <uses-permission android:name="android.permission.INTERNET" />
+  <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
```
Nit: Add one blank line below this. 
`Option+Cmd+L` to reorganise everything in this file.

jamesxu0 replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/app/activity/ActivityComponent.kt:
```diff
@@ -50,12 +50,12 @@ interface ActivityComponent {
   fun inject(topicTestActivity: TopicTestActivity)
   fun inject(topicTestActivityForStory: TopicTestActivityForStory)
   fun inject(storyFragmentTestActivity: StoryFragmentTestActivity)
+  fun inject(explorationInjectionActivity: ExplorationInjectionActivity)
```
Nit: I think we all are following the alphabetical order in this.

jamesxu0 replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/app/player/audio/AudioButtonListener.kt:
```diff
@@ -0,0 +1,8 @@
+  fun hideAudioButton()
+  fun showAudioStreamingOn()
+  fun showAudioStreamingOff()
+}
```
EOF is missing.

Also, Add kDoc for this interface.

BenHenning replied:
& KDoc for the methods since there are multiple. We should document when each is expected to be called.

jamesxu0 replied:
Added KDocs.

rt4914 commented on app/src/main/java/org/oppia/app/player/audio/AudioFragment.kt:
```diff
@@ -66,13 +40,22 @@ class AudioFragment : InjectableFragment(), LanguageInterface {
+
   /** Used in data binding to know if user is touching SeekBar */
   fun getUserIsSeeking() = audioFragmentPresenter.userIsSeeking
-
```
Nit: Re-add this blank line.

jamesxu0 replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/app/player/audio/AudioFragmentInterface.kt:
```diff
@@ -0,0 +1,21 @@
+
+  /** Used to pause audio when hiding [AudioFragment]. */
+  fun pauseAudio()
+}
```
EOF is missing

jamesxu0 replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt:
```diff
@@ -1,66 +1,49 @@
 
 /** The presenter for [AudioFragment]. */
 @FragmentScope
 class AudioFragmentPresenter @Inject constructor(
```
Nit: Press `Option+Cmd+L` to reorganise everything.

jamesxu0 replied:
Organized file.

rt4914 commented on app/src/main/java/org/oppia/app/player/audio/LanguageInterface.kt:
```diff
@@ -4,4 +4,7 @@ package org.oppia.app.player.audio
   /** Play the audio corresponding to the language-selected. */
   fun onLanguageSelected(currentLanguageCode: String)
+
+  /** Open the language selection dialog */
```
Nit: `.` at the end of sentence.

jamesxu0 replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationViewModel.kt:
```diff
@@ -0,0 +1,14 @@
+  val showAudioButton = ObservableField(false)
+  /** Used to change audio button src. */
+  val isAudioStreamingOn = ObservableField(false)
+}
```
EOF is missing

jamesxu0 replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -247,28 +266,42 @@ class StateFragmentPresenter @Inject constructor(
+            }
+          }
+        }
+        ConnectionStatus.NONE-> {
```
Nit: `ConnectionStatus.NONE -> {`

jamesxu0 replied:
Updated.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/player/audio/AudioFragmentTest.kt:
```diff
@@ -295,6 +284,10 @@ class AudioFragmentTest {
     @Provides
     fun provideGlobalLogLevel(): LogLevel = LogLevel.VERBOSE
+
+    @CacheAssetsLocally
```
<img width="921" alt="Screenshot 2019-11-28 at 11 59 17 PM" src="https://user-images.githubusercontent.com/9396084/69826902-2157e400-123b-11ea-9cf7-e49c48857f49.png">

Was not able run these test because of this error.

BenHenning replied:
Merge error?

jamesxu0 replied:
Removed duplicate @Provides.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -21,37 +25,206 @@ import dagger.Component
 
 /** Tests for [ExplorationActivity]. */
 @RunWith(AndroidJUnit4::class)
 class ExplorationActivityTest {
```
<img width="1638" alt="Screenshot 2019-11-29 at 12 02 46 AM" src="https://user-images.githubusercontent.com/9396084/69827009-9cb99580-123b-11ea-874e-9fcb95231ada.png">

These test cases are failing and all for same reason.

jamesxu0 replied:
All tests pass when I run them on a physical device. I have the same tests failing when running on an emulator. Not sure how to resolve this. Any suggestions @BenHenning?

BenHenning replied:
I don't think that picture shows the 'caused by'. What's the underlying issue?

rt4914 reviewed (commented):
Mostly nit changes, and issues with test-cases. Please have a look.

Also, I am not requesting changes, just so it does not block you.

BenHenning commented on app/src/main/AndroidManifest.xml:
```diff
@@ -15,20 +16,19 @@
     <activity
       android:name=".home.continueplaying.ContinuePlayingActivity"
-      android:theme="@style/OppiaThemeWithoutActionBar"
-      android:screenOrientation="portrait" />
```
Please re-add. We don't support landscape anywhere in the app currently.

jamesxu0 replied:
Re-added.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragmentInterface.kt:
```diff
@@ -0,0 +1,21 @@
+import androidx.lifecycle.LiveData
+
+/** Interface to interact with AudioFragment from [StateFragment]. */
+interface AudioFragmentInterface {
```
Maybe: AudioUiManager

And for the documentation, we could frame this as: ``Manager for updating audio state within the state player.``

We shouldn't tie this in documentation or name to AudioFragment since that's sort of an implementation detail. 'AudioFragmentManager' might be reasonable, as well, but I think this might be slightly cleaner. Thoughts?

jamesxu0 replied:
That makes sense to me. Updated to AudioUiManager.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragmentInterface.kt:
```diff
@@ -0,0 +1,21 @@
+interface AudioFragmentInterface {
+  /**
+   * Used to set AudioFragment's VoiceoverMappings with given stateId and contentId.
+   * If no contentId is provided, AudioFragment will use state.content.contentId as default.
```
This is a slightly complicated contract.  Could we just always pass in the correct content ID or does the underlying fragment definitely need to subscribe to the state to retrieve it? It seems simpler to me to pass in data where possible to minimize live data observers.

jamesxu0 replied:
Separated setVoiceoverMapping into two separate functions to reduce complexity.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt:
```diff
@@ -96,63 +72,38 @@ class AudioFragmentPresenter @Inject constructor(
-      viewModel.setAudioLanguageCode(selectedLanguageCode)
-    })
+  fun setVoiceoverMappings(explorationId: String, stateId: String, contentId: String? = null) = viewModel.setVoiceoverMappings(explorationId, stateId, contentId)
+  fun getCurrentPlayStatus() = viewModel.playStatusLiveData
```
Nit here & elsewhere: please make sure there's 1 empty newline between function declarations for spacing.

jamesxu0 replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -1,27 +1,45 @@
+  private lateinit var exploration: Exploration
   private lateinit var explorationId: String
   private var voiceoverMap = mapOf<String, Voiceover>()
+  private val defaultLanguage = "en"
```
Shouldn't this be Hindi post #442?

jamesxu0 replied:
Updated to default to "hi".

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -114,4 +169,13 @@ class AudioViewModel @Inject constructor(
     return "https://storage.googleapis.com/$gcsResource/exploration/$explorationId/assets/audio/${voiceover?.fileName}"
   }
+
+  private fun <T> LiveData<T>.observeOnce(lifecycleOwner: LifecycleOwner, observer: Observer<T>) {
```
Can we save this extension function in a Kotlin file somewhere rather than hiding it in a class? I think it's really useful and have found myself wanting other times.

jamesxu0 replied:
Moved to separate extensions package.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -22,17 +25,28 @@ private const val TAG_EXPLORATION_FRAGMENT = "TAG_EXPLORATION_FRAGMENT"
 ) {
-
   private lateinit var toolbar: Toolbar
+  private val exploreViewModel by lazy {
```
Nit: ``explorationViewModel`` for consistency.

jamesxu0 replied:
I can't rename it to explorationViewModel because it has the same JVM signuature as getExplorationviewModel().

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationViewModel.kt:
```diff
@@ -0,0 +1,14 @@
+import javax.inject.Inject
+
+@ActivityScope
+class ExplorationViewModel @Inject constructor(): ObservableViewModel() {
```
Please add KDoc.

jamesxu0 replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -123,12 +138,15 @@ class StateFragmentPresenter @Inject constructor(
       it.viewModel = this.viewModel
     }
 
+    // Initialize Audio Player
```
Should we only do this if AudioFragment isn't already added (such as after a configuration change)?

jamesxu0 replied:
Added null check to not add duplicate fragments.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -377,7 +456,7 @@ class StateFragmentPresenter @Inject constructor(
   }
 
-  /**
+  /**clickNegative_checkNo
```
Remove?

jamesxu0 replied:
Removed.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -405,6 +490,14 @@ class StateFragmentPresenter @Inject constructor(
 
+  private fun showOrHideAudioByState(state: State) {
+    if (state.recordedVoiceoversCount == 0) {
+      (fragment.requireActivity() as AudioButtonListener).hideAudioButton()
```
Could we use the injected activity instead of calling requireActivity?

jamesxu0 replied:
updated to use injected activity.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -572,4 +668,15 @@ class StateFragmentPresenter @Inject constructor(
+      .setMessage(context.getString(R.string.audio_dialog_offline_message))
+      .setPositiveButton(context.getString(R.string.audio_dialog_offline_positive)) { dialog, _ ->
+        dialog.dismiss()
+      }.create().show()
```
We should show it immediately rather than asynchronously.

jamesxu0 replied:
I'm not sure how to sure immediately rather than asynchronously. I can only use the show method on the dialogfragment.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -572,4 +668,15 @@ class StateFragmentPresenter @Inject constructor(
     inputManager.hideSoftInputFromWindow(fragment.view!!.windowToken, InputMethodManager.SHOW_FORCED)
   }
+
+  private fun showOfflineDialog() {
```
For future work: it seems a bit odd to me that state fragment is responsible for handling all of this audio state. I can see it needing to provide the necessary IDs, handle play, and deal with auto-play scenarios. However, managing when to show the dialog or offline/cellular notices should probably be handled elsewhere, maybe in an AudioManagerFragment.

jamesxu0 replied:
Moved all logic related to audio to AudioFragment.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -294,28 +327,67 @@ class StateFragmentPresenter @Inject constructor(
+      if (it == AudioViewModel.UiAudioPlayStatus.COMPLETED) {
+        getAudioFragment()?.let {
+          if (isFeedbackPlaying) {
+            audioFragmentInterface.setVoiceoverMappings(explorationId, currentStateName)
```
Should we be passing feedbackId here?

jamesxu0 replied:
Comment is out of date because API changed.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -294,28 +327,67 @@ class StateFragmentPresenter @Inject constructor(
+      binding.stateRecyclerView.smoothScrollToPosition(0)
+    }
+    getAudioFragment()?.let {
+      (it as AudioFragmentInterface).setVoiceoverMappings(explorationId, currentStateName, feedbackId)
```
In general, I'm having a lot of difficulty following the logic here. We are setting voiceover mappings in a bunch of places--can that be simplified? E.g. it would be ideal if we just set all voiceover mappings for the state once, then tell the audio manager to play certain content IDs at certain times. As it stands, I'm really concerned that changing some of this logic around will have subtle breakages that are hard to track down. We should try to make the implementation robust by isolating important arrangement code to just one place, and lean on special cases to properly change state (e.g. playing feedback audio should only involve operating with something related to feedback, like a content ID, rather than setting a bunch of state to make it work).

Happy to discuss this further if needed.

jamesxu0 replied:
Changed API of AudioUiManager to have setStateAndExplorationId and loadAudio. This simplified the logic inside StateFragment so that it only has to update state for the AudioFragment and load audio when it needs to based on content id.

BenHenning commented on app/src/main/res/drawable/ic_audio_streaming_off_24dp.xml:
```diff
@@ -0,0 +1,5 @@
+<vector android:autoMirrored="true" android:height="24dp"
```
As I understand it, we should only flip icons when they imply directionality. Otherwise, we should keep the icon the same in RTL and LTR layouts. I don't think the speaker icon falls under the directionality case, so I suggest not auto-mirroring it.

Ditto for the 'on' version.

jamesxu0 replied:
Removed autoMirrored.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -21,37 +25,206 @@ import dagger.Component
+  }
+
+  @Test
+  fun testAudioWithCellular_openRatioExploration_clickCheckboxAndPositive_clickAudioIconTwice_checkAudioFragmentIsVisibleAndDialogIsNotDisplayed() {
```
For here & above & below: please make sure the arrange/act/assert blocks are clearly indicated to better help with the test readability. This is less important for shorter tests, but these tests are getting long enough that it's starting to get hard to follow.

jamesxu0 replied:
Added spacing to follow arrange/act/assert.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -21,37 +25,206 @@ import dagger.Component
+      onView(withId(R.id.text_input_interaction_view)).perform(ViewActions.typeText("123"), closeSoftKeyboard())
+      onView(withId(R.id.interaction_button)).perform(click())
+      // Ensures that feedback audio starts playing
+      Thread.sleep(1000)
```
This may result in a flaky test. Is there a more reliable wait that we can do?

jamesxu0 replied:
I'm not sure how to wait for the audio to start playing without sleep. Any suggestions?

BenHenning replied:
Is this actually downloading audio, then? If we're using the real media player, we should probably ``@Ignore`` this test & related until we can actually swap out with a fake implementation.

jamesxu0 replied:
Updated to ignore test.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -21,37 +25,206 @@ import dagger.Component
+    getApplicationDependencies(RATIOS_EXPLORATION_ID_0)
+    networkConnectionUtil.setCurrentConnectionStatus(NetworkConnectionUtil.ConnectionStatus.LOCAL)
+    ActivityScenario.launch<ExplorationActivity>(createExplorationActivityIntent(RATIOS_EXPLORATION_ID_0)).use {
+      for (x in 0..4) {
```
Suggest duplicating the line. It's generally bad practice to have control flow in tests; prefer code duplication, instead. I also suggest adding a brief comment to indicate why we're intentionally clicking the same button multiple times.

jamesxu0 replied:
Updated.

BenHenning commented on domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt:
```diff
@@ -118,14 +119,21 @@ class AudioPlayerController @Inject constructor(
+    mediaPlayer.setOnErrorListener {_, _, _ ->
+      releaseMediaPlayer()
+      initializeMediaPlayer()
+      true // Indicates that error was handled and to not invoke completion listener.
```
Suggesting using a return@ here to make it a bit clearer.

BenHenning replied:
Also, should this trigger an error AsyncResult in playProgress?

jamesxu0 replied:
Updated return@ and sets playProgress to failed AsyncResult.

BenHenning commented on domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt:
```diff
@@ -349,14 +349,17 @@ class AudioPlayerControllerTest {
-  fun testController_alreadyInitialized_initializePlayer_fails() {
-    arrangeMediaPlayer()
+  @ExperimentalCoroutinesApi
+  fun testController_invokeErrorListener_invokePrepared_verifyNoNewStatusUpdates()
```
Please make the test name more explicit as to what the updated expected status is.

jamesxu0 replied:
updated to verifyAudioPlayerIsStillPending.

BenHenning commented on utility/src/main/java/org/oppia/util/networking/NetworkConnectionUtil.kt:
```diff
@@ -0,0 +1,41 @@
+@Singleton
+class NetworkConnectionUtil @Inject constructor(private val context: Context) {
+  enum class ConnectionStatus {
+    LOCAL, // Connected to WIFI or Ethernet
```
Nit: this should be a KDoc on the constant rather than a single line comment. Ditto for others.

jamesxu0 replied:
Updated.

BenHenning reviewed (changes requested):
In general, I think most of this PR looks good. My main concern is the complexity of StateFragmentPresenter is going up significantly, and that's already a quite complex class. We should aim to keep changes the simplest they can be to help with future changes to the presenter.

I think once the presenter is handled a bit more cleanly, this PR will LGTM.

jamesxu0 committed:
Minor fixes

jamesxu0 committed:
Merge branch 'develop' into state-fragment-audio

jamesxu0 committed:
Changed AudioUIManger API

jamesxu0 committed:
Moving code to audioFragmentPresenter (WIP).

jamesxu0 committed:
Removed audio code from statefragment

jamesxu0 committed:
Added annotations to test

jamesxu0 commented:
@BenHenning PTAL. I moved a lot of the logic out of StateFragment as we discussed before, so the AudioFragment should be reusable in QuestionPlayerFragment. I simplified the API of AudioUiManager to have separate functions for setting the current state and loading audio by content id.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt:
```diff
@@ -96,63 +105,129 @@ class AudioFragmentPresenter @Inject constructor(
+  private fun showAudioFragment() {
+    getStateFragment()?.setAudioBarVisibility(true)
+    (activity as AudioButtonListener).showAudioStreamingOn()
+    loadAudio(feedbackId, true)
```
Nit: please add argument name in-line comment for clarity.

jamesxu0 replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioFragmentPresenter.kt:
```diff
@@ -96,63 +105,129 @@ class AudioFragmentPresenter @Inject constructor(
+    val animation = AnimationUtils.loadAnimation(context, R.anim.slide_up_audio)
+    animation.setAnimationListener(object : Animation.AnimationListener {
+      override fun onAnimationEnd(p0: Animation?) {
+        getStateFragment()?.setAudioBarVisibility(false)
```
Here & elsewhere: we should't put a dependency on StateFragment. Could we indicate audio bar visibility using a listener pattern again that goes up through the activity?

jamesxu0 replied:
Updated to route function through the activity.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioUiManager.kt:
```diff
@@ -0,0 +1,25 @@
+   */
+  fun loadAudio(contentId: String?, allowAutoPlay: Boolean)
+
+  /** Used to pause audio when hiding [AudioFragment]. */
```
Is the 'when hiding fragment' bit important? Could we just say this pauses all playing audio?

jamesxu0 replied:
Updated to "pause all playing audio".

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioUiManager.kt:
```diff
@@ -0,0 +1,25 @@
+   * Loads audio for media player with content id.
+   *
+   * @param contentId: Used to get a specific Voiceover. If null, state.content.contentId is used as default.
+   * @param allowAutoPlay: If false, audio is guaranteed not to be autoPlayed.
```
Should we make this stricter? E.g. if this is false, it doesn't ever autoplay? And, if so, when would we use this? There doesn't seem to be another case where audio can start playing.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -1,30 +1,49 @@
+  private lateinit var state: State
   private lateinit var explorationId: String
   private var voiceoverMap = mapOf<String, Voiceover>()
+  private val defaultLanguage = "hi"
```
Should be "en" I believe. We just switched this to "hi" for the temp branch.

jamesxu0 replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -44,16 +63,37 @@ class AudioViewModel @Inject constructor(
+      languages.isNotEmpty() -> {
+        autoPlay = false
+        languageSelectionShown = true
+        val languageCode = if (languages.any { it == "en" }) {
```
Can we use contains? E.g. if (languages.contains("en")) {...}

jamesxu0 replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/app/player/audio/AudioViewModel.kt:
```diff
@@ -103,11 +143,20 @@ class AudioViewModel @Inject constructor(
       PlayStatus.PAUSED -> UiAudioPlayStatus.PAUSED
-      PlayStatus.COMPLETED -> UiAudioPlayStatus.COMPLETED
+      PlayStatus.COMPLETED -> {
+        if (hasFeedback) loadAudio(null, false)
```
What does this do? If it's for deinitialization, could we do that in a private helper? I'm generally not a fan of using a method for something that seems unrelated and only is triggered by sentinel values--I find that hard to understand in context, which itself can lead to future bugs if other developers misinterpret what this is intending to do.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationViewModel.kt:
```diff
@@ -0,0 +1,16 @@
+  /** Used to control visibility of audio button. */
+  val showAudioButton = ObservableField(false)
+
+  /** Used to change audio button src. */
```
How about to change the audio button enabled status?

jamesxu0 replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -173,27 +177,37 @@ class StateFragmentPresenter @Inject constructor(
-          ContentItemBinding.inflate(LayoutInflater.from(parent.context), parent, /* attachToParent= */ false).root
+          ContentItemBinding.inflate(
+            LayoutInflater.from(parent.context),
+            parent, /* attachToParent= */
```
Nit here & below: move comment to next line to be next to parameter.

jamesxu0 replied:
Updated.

BenHenning commented on app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -1,19 +1,17 @@
 
-import android.animation.Animator
-import android.animation.AnimatorListenerAdapter
+import android.app.AlertDialog
```
Do we still need this import? Please remove unused imports, if any.

jamesxu0 replied:
Removed unused imports.

BenHenning commented on utility/src/main/java/org/oppia/util/extensions/LiveDataExt.kt:
```diff
@@ -0,0 +1,14 @@
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.Observer
+
+fun <T> LiveData<T>.observeOnce(lifecycleOwner: LifecycleOwner, observer: Observer<T>) {
```
Please add KDoc explaining what this does & how to use it.

BenHenning replied:
Also, thanks for adding this. :D

jamesxu0 replied:
Updated.

BenHenning commented on utility/src/main/java/org/oppia/util/extensions/LiveDataExt.kt:
```diff
@@ -0,0 +1,14 @@
+      removeObserver(this)
+    }
+  })
+}
```
EOF newline.

jamesxu0 replied:
Updated.

BenHenning reviewed (approved):
Overall, I think this is an excellent simplification of the state fragment presenter. LGTM. Let me know if you want to discuss some of the architecture or open ended questions.

Thanks!

jamesxu0 committed:
Minor changes

jamesxu0 committed:
Merge branch 'develop' into state-fragment-audio

jamesxu0 committed:
Merge branch 'develop' into state-fragment-audio

jamesxu0 committed:
Routed setVisibility to activity

jamesxu0 committed:
Added scrollToTop.

jamesxu0 committed:
Updated AudioUiManager
